aiida/tools/importexport/dbexport/utils.py

```diff
@@ -403,171 +402,39 @@ def retrieve_linked_nodes(process_nodes, data_nodes, **kwargs):  # pylint: disab
-                continue
-            # Otherwise say that it is a node to be exported
-            else:
-                retrieved_nodes.add(current_node_pk)
-
-            # INPUT_CALC(Data, CalculationNode) - Forward
-            if traversal_rules['input_calc_forward']:
-                links_uuids, found_nodes = _retrieve_linked_nodes_query(
-                    current_node_pk,
-                    input_type=Data,
-                    output_type=ProcessNode,
-                    direction='forward',
-                    link_type_value=LinkType.INPUT_CALC.value
-                )
-                process_nodes.update(found_nodes - retrieved_nodes)
-                links_uuid_dict.update(links_uuids)
-
-            # CREATE(CalculationNode, Data) - Backward
-            if traversal_rules['create_backward']:
-                links_uuids, found_nodes = _retrieve_linked_nodes_query(
-                    current_node_pk,
-                    input_type=ProcessNode,
-                    output_type=Data,
-                    direction='backward',
-                    link_type_value=LinkType.CREATE.value
-                )
-                process_nodes.update(found_nodes - retrieved_nodes)
-                links_uuid_dict.update(links_uuids)
-
-            # RETURN(WorkflowNode, Data) - Backward
-            if traversal_rules['return_backward']:
-                links_uuids, found_nodes = _retrieve_linked_nodes_query(
-                    current_node_pk,
-                    input_type=ProcessNode,
-                    output_type=Data,
-                    direction='backward',
-                    link_type_value=LinkType.RETURN.value
-                )
-                process_nodes.update(found_nodes - retrieved_nodes)
-                links_uuid_dict.update(links_uuids)
-
-            # INPUT_WORK(Data, WorkflowNode) - Forward
-            if traversal_rules['input_work_forward']:
-                links_uuids, found_nodes = _retrieve_linked_nodes_query(
-                    current_node_pk,
-                    input_type=Data,
-                    output_type=ProcessNode,
-                    direction='forward',
-                    link_type_value=LinkType.INPUT_WORK.value
-                )
-                process_nodes.update(found_nodes - retrieved_nodes)
-                links_uuid_dict.update(links_uuids)
-
-    return retrieved_nodes, list(links_uuid_dict.values()), traversal_rules
+    # Creating a set of pks that are the process nodes and data nodes:
+    starting_nodes = process_nodes.union(data_nodes)
+
+    # Calling the traverser, which will traverse the graph exhaustively until it converges to
+    # a stable subgraph
+    traverse_results = traverse_graph(starting_nodes, get_links=True, **traversal_rules)
+
+    # The exporter wants a different form, a list of dictionaries for every link.
+    # I create a utility dictionary for pk to uuid. In the future, this should be
+    # ideally done in the interface or AGE.
```

Comment:
should one open an issue for this?

Reply:
There might be a couple of 'TO-DO' sentences scattered around, I can gather all of this and remove them from the code and open some issues if that is better, the only thing is that these might stay 'orphaned' for a while.

Reply:
yes, please do.

> the only thing is that these might stay 'orphaned' for a while.

well, that's not your problem right now ;-)